Cython/Utility/Builtins.c

```diff
@@ -22,8 +22,13 @@ static PyObject* __Pyx_Globals(void); /*proto*/
 static PyObject* __Pyx_Globals(void) {
     Py_ssize_t i;
     PyObject *names;
+#if CYTHON_COMPILING_IN_LIMITED_API
+    PyObject *globals = PyDict_New();
```

Comment:
Let me leave this as unresolved. I'll get back to you on this

Reply:
ok

Reply:
Actually, I've been looking at this code for a long time and I have a couple of questions:

Why do we call `PyObject_Dir` if we already have the module dictionary? Internally `PyObject_Dir` will call `moduleobject.c:module_dir` which ends up calling `PyDict_Keys`. This will exactly match the keys from the module dict that we already have. At least this is what happens in CPython master. Maybe some older version did not do the same?

Thus, what about changing this to:
```
static PyObject* __Pyx_Globals(void) {
    PyObject *globals;
#if CYTHON_COMPILING_IN_LIMITED_API
    globals = PyModule_GetDict($module_cname);
    if (unlikely(!globals)) return NULL;
#else
    globals = $moddict_cname;
#endif
    Py_INCREF(globals);
    return globals;
}
```

I could even add a separate PR to simplify this `__Pyx_Globals `

Reply:
`PyObject_Dir()` used to do a lot of different things in Py2, but looking at the implementation in Py2.7 now, there's a dedicated path for modules, which just returns the keys of the module dict, as you say. I think the proposed simplification is ok.

Also, the comment above the implementation is outdated, but gives a hint why the implementation might have ended up like this. Since supporting `cdef` globals is not implemented currently, I think it's safe to simplify the whole thing and remove the comment along the way. Seems worth a separate little PR.

(Careful with comments like "All tests are passing with the proposed change", BTW. Many things simply weren't worth testing in the past, and would have let the test suite explode if we had exhaustively tested the Python-compatible behaviour. Also, behavioural tests for Python language features that we (re-)implemented were not always available in the test suite of CPython itself, so we couldn't just have copied them over or so. That's only slowly changing over there in recent years, also triggered by the PyPy project and other implementations.)

Reply:
(Ah, I hadn't seen that you already changed this – don't need a new PR then.)